### DISCUSSION TITLE
API Naming Convention API 命名统一

### DISCUSSION THREAD

**Discussion by kooriookami** [element-plus/element-plus#16291] [RFC]
## API Naming Convention API 命名统一

API naming should follow the principle of simplicity and clarity. A list for the naming inconsistencies in the components is open for discussion.

API 命名应该遵循简洁明了的原则。针对组件中命名不统一的地方进行列举，欢迎大家讨论。

### Component Attributes  组件属性

The same functionality across different components should be consistent.

不同的组件相同的功能应该保持一致。

| Current 当前 | Convention 统一 | Discription 描述 |
| --- | --- | --- |
| show-after / show-timeout | show-after | delay time before show, please use `packages/hooks/use-delayed-toggle` <br> 延迟显示，请使用 `packages/hooks/use-delayed-toggle` |
| hide-after / hide-timeout | hide-after | delay time before hide, please use `packages/hooks/use-delayed-toggle` <br> 延迟关闭，请使用 `packages/hooks/use-delayed-toggle` |

### Naive Attributes 原生属性

Naive attributes of tags names should be consistent.

标签原生属性名称应该保持一致。

| Current 当前 | Convention 统一 | Discription 描述 |
| --- | --- | --- |
| label / aria-label | aria-label | native `aria-label` attribute <br> 原生 `aria-label` 属性 |

**Comment by xxholly32** [comment]
add my optionion about css-var RFC:  https://github.com/element-plus/element-plus/discussions/16709

**Comment by xxholly32** [comment]
tag 有3个api有点奇怪

+ effect 原本 用在 tooltip 里面还是可以理解，但在 tag 中却用了 effect: dark 实现了类似 button 的效果，但button中是使用bg，text，link去解决的(不过我认为button的属性也太多了)；是否2者可以考虑仅通过cssvar去解决
+ hit 感觉叫 bordered 可能比较形象~ (checkbox里面叫border)
+ `color` el-button 有个 color 属性在 beta 可以生成颜色了，而 tag 却没有

It's a little weird that tag 3 apis

+ `effect` is still understandable when used in `el-tooltip`, but `effect: dark` is used in `el-tag` to achieve a button-like effect, but the button is solved using `bg`, `text`, and `link` (But I think button has too many attributes)
+ `hit` feel called bordered may compare image ~
+ `color` `el-button` has a beta that generates colors automatically, whereas `el-tag` does not

**Comment by baixiaoyu2997** [comment]
我注意到一个有趣的问题, table v2中对于事件放弃了emits的写法，转而使用props（naive-ui也是这么处理的），我猜测这是因为defineEmits的一些局限性， 但是在最新的组件中却还是在使用emits, 我想问一下，开发团队是否会在未来都替换为使用props？
![image](https://github.com/element-plus/element-plus/assets/9149999/a7b230e9-098c-45f8-b58f-322f5633b079)

![image](https://github.com/element-plus/element-plus/assets/9149999/addeeeb4-ef04-4a1a-8eaf-5f4a6dc78b0d)

**Comment by xxholly32** [comment]
有关 inject/provider 的 key，没有统一

存在几种情况

+ 不带 el 的（常见）
+ 带 el 的 （常见）
+ 带 ep 的 （不常见）https://github.com/element-plus/element-plus/blob/dev/packages/components/time-picker/src/common/picker.vue#L758
+  啥都不带~ （不常见） https://github.com/element-plus/element-plus/blob/dev/packages/components/date-picker/src/constants.ts#L9

**Comment by nasirovx** [comment]
I agree with the proposal to use `show-after` instead of `show-timeout` and `aria-label` instead of `label`. This will improve naming consistency and make working with the API easier.

I'd also like to suggest the following changes:
- In the `el-button` component, consider using `bg-color` instead of `effect` to maintain uniformity.
- For the `el-input` component, use `max-length` instead of `maxlength` to align with standard conventions.

What do other contributors think about these suggestions?
